[bbc_sale][ADD] Introduce is_synced_magento flag on product to search on synced to Magento products

wytzejan opened the pull request:
is_synced_magento boolean introduced on product.template en product.product form.
It is a computed boolean field: if product_id is available in magento.product.template the boolean is yes; otherwise no. For product.product it looks to the model magento.product.
The field is available in the advanced search option of the product form. The search-function is called the moment we search on the field. The purpose of the field is that we would like to be able to search on the synced/not synced to Magento products.

wytzejan committed:
[bbc_sale][ADD] Introduce is_synced_magento flag on product to search on synced to Magento products

wytzejan committed:
[bbc_sale][ADD] Introduce house_number and house_number_addition to partner form

wytzejan committed:
Revert "[bbc_sale][ADD] Introduce house_number and house_number_addition to partner form"

This reverts commit a69edbb0b43e2dfb29fc39517d2470649ad29642.

StefanRijnhart commented on bbc_sale/models/product.py:
```diff
@@ -67,6 +71,23 @@ def search_is_component(self, operator, value):
+        if operator in ('!=', '<>'):
+            negate = not negate
+        templates = self.env['magento.product.template'].search(
+            []).mapped('erp_template_id')
```
This should be slightly faster:

```
templates = self.env['magento.product.template'].search([
    ('erp_template_id', 'not in', (False, 0))])
return [('id', 'not in' if negate else 'in', templates.ids)]
```

StefanRijnhart commented on bbc_sale/models/product.py:
```diff
@@ -67,6 +71,23 @@ def search_is_component(self, operator, value):
         return [('id', 'not in' if negate else 'in', templates.ids)]
 
     @api.multi
+    def compute_is_synced_magento(self):
```
perform the search outside of the loop for better performance:
```
synced = self.env['magento.product.template'].search(
    [('erp_template_id', 'in', template.ids)]).mapped('erp_template_id')
for template in self:
    template.is_synced_magento = template.id in synced
```

wytzejan replied:
Fixed.

StefanRijnhart commented on bbc_sale/models/product.py:
```diff
@@ -239,6 +264,23 @@ def search_is_component(self, operator, value):
     @api.multi
+    def compute_is_synced_magento(self):
+        for product in self:
+            product.is_synced_magento = bool(
```
same as above

StefanRijnhart commented on bbc_sale/models/product.py:
```diff
@@ -239,6 +264,23 @@ def search_is_component(self, operator, value):
+        if operator in ('!=', '<>'):
+            negate = not negate
+        products = self.env['magento.product'].search(
+            []).mapped('oe_product_id')
```
same as above

StefanRijnhart commented on bbc_sale/views/product.xml:
```diff
@@ -146,5 +146,17 @@
             </field>
         </record>
 
+        <record id="website_sale_product_template_form_view_inherited_is_synced_magento" model="ir.ui.view">
```
Can this be added to the existing product_template_form_view instead of adding a new view definition? The existing view already has a very 'high' priority so it is bound to already include the override from website_sale.

wytzejan committed:
[bbc_sale][ADD] changed functions and views according to reviews

wytzejan commented:
@StefanRijnhart 
Regarding computed functions: I moved the search outside the loop.

Regarding search functions: I was not able to copy-paste your proposal. Behind the search function I added the .mapped('erp_template_id')-function (for product.product: .mapped('oe_product_id')).
This is needed because the search-function must compare the template or product id with the erp_template_id/oe_product_id instead of the ID of the record in magento.product.template/magento.product.

Regarding view: I moved the field is_synced_magento to the existing product_template_form_view.

StefanRijnhart commented:
Indeed, I made a mistake but you got the idea. Thanks for the updates!
